rclrs/src/node/graph.rs

```diff
@@ -0,0 +1,635 @@
+pub type TopicNamesAndTypes = HashMap<String, Vec<String>>;
+
+/// Stores a node's name and namespace
+#[derive(Debug, PartialEq)]
+pub struct NodeNameInfo {
+    /// The name of the node
+    pub name: String,
+    /// The namespace of the node
+    pub namespace: String,
+}
+
+/// Contains topic endpoint information
+pub struct TopicEndpointInfo {
+    /// The name of the endpoint node
+    pub node_name: String,
+    /// The namespace of the endpoint node
+    pub node_namespace: String,
+    /// The type of the topic
+    pub topic_type: String,
+    /// The RMW endpoint type of the topic
+    pub endpoint_type: rmw_endpoint_type_t,
+}
+
+impl Node {
+    /// Get a list of topic names and types for publishers associated with a node.
+    pub fn get_publisher_names_and_types_by_node(
+        &self,
+        node: &str,
+        namespace: &str,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        self.get_names_and_types_by_node_with_demangle_arg(
+            node,
+            namespace,
+            rcl_get_publisher_names_and_types_by_node,
+        )
+    }
+
+    /// Get a list of topic names and types for subscriptions associated with a node.
+    pub fn get_subscription_names_and_types_by_node(
+        &self,
+        node: &str,
+        namespace: &str,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        self.get_names_and_types_by_node_with_demangle_arg(
+            node,
+            namespace,
+            rcl_get_subscriber_names_and_types_by_node,
+        )
+    }
+
+    /// Get a list of topic names and types for services associated with a node.
+    pub fn get_service_names_and_types_by_node(
+        &self,
+        node: &str,
+        namespace: &str,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        self.get_names_and_types_by_node_without_demangle_arg(
+            node,
+            namespace,
+            rcl_get_service_names_and_types_by_node,
+        )
+    }
+
+    /// Get a list of topic names and types for clients associated with a node.
+    pub fn get_client_names_and_types_by_node(
+        &self,
+        node: &str,
+        namespace: &str,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        self.get_names_and_types_by_node_without_demangle_arg(
+            node,
+            namespace,
+            rcl_get_client_names_and_types_by_node,
+        )
+    }
+
+    /// Get a list of all topic names and their types.
+    pub fn get_topic_names_and_types(&self) -> Result<TopicNamesAndTypes, RclrsError> {
+        let mut rcl_names_and_types = unsafe { rmw_get_zero_initialized_names_and_types() };
+
+        // SAFETY: rcl_names_and_types is zero-initialized as expected by this call
+        unsafe {
+            rcl_get_topic_names_and_types(
+                &*self.rcl_node_mtx.lock().unwrap() as *const _,
+                &mut rcutils_get_default_allocator() as *mut _,
+                false,
+                &mut rcl_names_and_types as *mut _,
```

Comment:
Casting references to raw pointers should not be needed, and we generally avoid it.

```suggestion
                &*self.rcl_node_mtx.lock().unwrap(),
                &mut rcutils_get_default_allocator(),
                false,
                &mut rcl_names_and_types,
```